pkg/controller/versions/template.go

```diff
@@ -69,58 +70,89 @@ func mergeValues(base map[string]interface{}, input map[string]interface{}) map[
+		return nil, nil, fmt.Errorf("profile name contains invalid character '/'")
 	}
 
-	templateContent, err := ioutil.ReadFile(path.Join(v.GetUserTemplatesDir(), name))
+	profileContent, err := ioutil.ReadFile(path.Join(v.GetUserTemplatesDir(), name))
 	if err != nil {
-		//if we can't read from the user template path, try from the default path
+		//if we can't read from the user profile path, try from the default path
 		//we use two paths because Kubernetes will not auto-update volume mounted
 		//configmaps mounted in directories with pre-existing content
-		defaultTemplateContent, defaultErr := ioutil.ReadFile(path.Join(v.GetDefaultTemplatesDir(), name))
+		defaultProfileContent, defaultErr := ioutil.ReadFile(path.Join(v.GetDefaultTemplatesDir(), name))
 		if defaultErr != nil {
-			return v1.ControlPlaneSpec{}, fmt.Errorf("template cannot be loaded from user or default directory. Error from user: %s. Error from default: %s", err, defaultErr)
+			return nil, nil, fmt.Errorf("template cannot be loaded from user or default directory. Error from user: %s. Error from default: %s", err, defaultErr)
 		}
-		templateContent = defaultTemplateContent
+		profileContent = defaultProfileContent
 	}
 
-	var template v1.ServiceMeshControlPlane
-	if err = yaml.Unmarshal(templateContent, &template); err != nil {
-		return v1.ControlPlaneSpec{}, fmt.Errorf("failed to parse template %s contents: %s", name, err)
+	obj, gvk, err := decoder.Decode(profileContent, nil, nil)
+	if err != nil || gvk == nil {
+		return nil, nil, fmt.Errorf("failed to parse profile %s contents: %s", name, err)
+	}
+	switch smcp := obj.(type) {
+	case *v1.ServiceMeshControlPlane:
+		if len(smcp.Spec.Profiles) == 0 {
+			if smcp.Spec.Template == "" {
+				return &smcp.Spec, nil, nil
+			} else {
+				return &smcp.Spec, []string{smcp.Spec.Template}, nil
+			}
+		}
+		return &smcp.Spec, smcp.Spec.Profiles, nil
+	case *v2.ServiceMeshControlPlane:
+		smcpv1 := &v1.ServiceMeshControlPlane{}
+		if err := smcpv1.ConvertFrom(smcp); err != nil {
+			return nil, nil, err
+		}
+		return &smcpv1.Spec, smcp.Spec.Profiles, nil
+	default:
+		return nil, nil, fmt.Errorf("unsupported ServiceMeshControlPlane version: %s", gvk.String())
 	}
-	return template.Spec, nil
 }
 
 //renderSMCPTemplates traverses and processes all of the references templates
-func (v version) recursivelyApplyTemplates(ctx context.Context, smcp v1.ControlPlaneSpec, visited sets.String) (v1.ControlPlaneSpec, error) {
+func (v version) recursivelyApplyProfiles(ctx context.Context, smcp *v1.ControlPlaneSpec, profiles []string, visited sets.String) (v1.ControlPlaneSpec, error) {
 	log := common.LogFromContext(ctx)
-	if smcp.Template == "" {
-		return smcp, nil
-	}
-	log.Info(fmt.Sprintf("processing smcp template %s", smcp.Template))
 
-	if visited.Has(smcp.Template) {
-		return smcp, fmt.Errorf("SMCP templates form cyclic dependency. Cannot proceed")
```

Comment:
I don't think this should be removed. It's really low hanging validation and the presence of circular dependencies means the user configured something improperly. Ignoring it it means that the user will not be achieving the behavior that they specified. We should find a way to surface this as a warning to the user before they encounter unrelated issues.

Reply:
Yeah, I guess I disagree.  They are getting the behavior they specified.  They just specified something that could break the operator (because we'd continually recurse).  The other problem is that because profiles is now a list, it's possible that a profile could be specified multiple times, without actually being circular.  All this does is break the chain, returning a result.  If the user doesn't get the result they wanted, then the configuration error is their problem, which isnot really any different than if they had specified bogus stuff in a single profile, that wasn't part of a circular chain.

I think the ultimate solution is to incorporate the profiles during the validation phase, so that we reject any invalid smcp before it gets to reconcile.  We should also add the resulting resource to the status, so users can inspect what they've got.  Adding a history would be nice too, but a better option might be to have some sort of tool that users can test with, or have some sort of dry-run capabilities.  Feel free to open a jira.

Reply:
yeah. Adding the resulting resource to the status does allow our users to go back when they start seeing problems and see what the result after parsing is. And adding the SMCP templates to validation would also help.